File: House_Robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        l=len(nums)
        
        if l==0:
            return 0
        
        t=[[0 for _ in range(2)]for _ in range(l)]
        t[0][1]=nums[0]
        
        for i in range(1,l):
            t[i][0]=max(t[i-1][0],t[i-1][1])
            t[i][1]=t[i-1][0]+nums[i]
            
        return max(t[l-1][0],t[l-1][1])
        

File: houseRobber.java
//DP
//TC: O(n)
//SC: O(n)
class Solution {
    public int rob(int[] nums) {
        if(nums == null || nums.length == 0) return 0;
        int n = nums.length;
        int[] dp = new int[n];
         if(n == 1) return nums[0];
        dp[0] = nums[0];
        dp[1] = Math.max(nums[0], nums[1]);
       
        for(int i=2; i< n; i++){
            dp[i] = Math.max(dp[i-1], dp[i-2]+nums[i]);
        }
        return dp[n-1];
    }
}


//DP
//TC: O(n)
//SC: O(1)
class Solution {
    public int rob(int[] nums) {
        if(nums == null || nums.length == 0) return 0;
        int n = nums.length;
        // int[] dp = new int[n];
        if(n == 1) return nums[0];
        int prev = nums[0];
        int curr = Math.max(nums[0], nums[1]);
      
       
        for(int i=2; i< n; i++){
            int temp = curr;
            curr = Math.max(curr,prev+nums[i]);
            prev = temp;
        }
        return curr;
    }
}
